// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: player/v1/plunder.proto

package dbv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on UserPlunderListProto with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UserPlunderListProto) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UserPlunderListProto with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UserPlunderListProtoMultiError, or nil if none found.
func (m *UserPlunderListProto) ValidateAll() error {
	return m.validate(true)
}

func (m *UserPlunderListProto) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	{
		sorted_keys := make([]int64, len(m.GetPlunders()))
		i := 0
		for key := range m.GetPlunders() {
			sorted_keys[i] = key
			i++
		}
		sort.Slice(sorted_keys, func(i, j int) bool { return sorted_keys[i] < sorted_keys[j] })
		for _, key := range sorted_keys {
			val := m.GetPlunders()[key]
			_ = val

			// no validation rules for Plunders[key]

			if all {
				switch v := interface{}(val).(type) {
				case interface{ ValidateAll() error }:
					if err := v.ValidateAll(); err != nil {
						errors = append(errors, UserPlunderListProtoValidationError{
							field:  fmt.Sprintf("Plunders[%v]", key),
							reason: "embedded message failed validation",
							cause:  err,
						})
					}
				case interface{ Validate() error }:
					if err := v.Validate(); err != nil {
						errors = append(errors, UserPlunderListProtoValidationError{
							field:  fmt.Sprintf("Plunders[%v]", key),
							reason: "embedded message failed validation",
							cause:  err,
						})
					}
				}
			} else if v, ok := interface{}(val).(interface{ Validate() error }); ok {
				if err := v.Validate(); err != nil {
					return UserPlunderListProtoValidationError{
						field:  fmt.Sprintf("Plunders[%v]", key),
						reason: "embedded message failed validation",
						cause:  err,
					}
				}
			}

		}
	}

	if len(errors) > 0 {
		return UserPlunderListProtoMultiError(errors)
	}

	return nil
}

// UserPlunderListProtoMultiError is an error wrapping multiple validation
// errors returned by UserPlunderListProto.ValidateAll() if the designated
// constraints aren't met.
type UserPlunderListProtoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UserPlunderListProtoMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UserPlunderListProtoMultiError) AllErrors() []error { return m }

// UserPlunderListProtoValidationError is the validation error returned by
// UserPlunderListProto.Validate if the designated constraints aren't met.
type UserPlunderListProtoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UserPlunderListProtoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UserPlunderListProtoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UserPlunderListProtoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UserPlunderListProtoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UserPlunderListProtoValidationError) ErrorName() string {
	return "UserPlunderListProtoValidationError"
}

// Error satisfies the builtin error interface
func (e UserPlunderListProtoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUserPlunderListProto.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UserPlunderListProtoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UserPlunderListProtoValidationError{}

// Validate checks the field values on UserPlunderProto with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *UserPlunderProto) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UserPlunderProto with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UserPlunderProtoMultiError, or nil if none found.
func (m *UserPlunderProto) ValidateAll() error {
	return m.validate(true)
}

func (m *UserPlunderProto) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Weights

	if len(errors) > 0 {
		return UserPlunderProtoMultiError(errors)
	}

	return nil
}

// UserPlunderProtoMultiError is an error wrapping multiple validation errors
// returned by UserPlunderProto.ValidateAll() if the designated constraints
// aren't met.
type UserPlunderProtoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UserPlunderProtoMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UserPlunderProtoMultiError) AllErrors() []error { return m }

// UserPlunderProtoValidationError is the validation error returned by
// UserPlunderProto.Validate if the designated constraints aren't met.
type UserPlunderProtoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UserPlunderProtoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UserPlunderProtoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UserPlunderProtoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UserPlunderProtoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UserPlunderProtoValidationError) ErrorName() string { return "UserPlunderProtoValidationError" }

// Error satisfies the builtin error interface
func (e UserPlunderProtoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUserPlunderProto.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UserPlunderProtoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UserPlunderProtoValidationError{}
